src/lib.rs

```diff
@@ -369,21 +369,21 @@ pub fn unfold<T, F, Fut, It, St>(init: T, f: F) -> Unfold<T, F, Fut>
 }
 
 /// A `StateStream` yielding elements of a normal `Stream.
-pub struct FromStream<S>(S);
+pub struct FromStream<S, T>(S, Option<T>);
 
-impl<S> StateStream for FromStream<S>
+impl<S, T> StateStream for FromStream<S, T>
     where S: Stream
 {
     type Item = S::Item;
-    type State = ();
+    type State = T;
     type Error = S::Error;
 
     #[inline]
-    fn poll(&mut self) -> Poll<StreamEvent<S::Item, ()>, S::Error> {
+    fn poll(&mut self) -> Poll<StreamEvent<S::Item, T>, S::Error> {
         self.0.poll().map(|a| {
             match a {
                 Async::Ready(Some(i)) => Async::Ready(StreamEvent::Next(i)),
-                Async::Ready(None) => Async::Ready(StreamEvent::Done(())),
+                Async::Ready(None) => Async::Ready(StreamEvent::Done(self.1.take().unwrap())),
```

Comment:
Could you replace the `unwrap` with an `expect("poll called after completion")` or something like that?